sled-agent/src/instance.rs

```diff
@@ -237,7 +357,92 @@ struct InstanceInner {
     instance_ticket: InstanceTicket,
 }
 
-impl InstanceInner {
+impl InstanceRunner {
+    async fn run(mut self) -> Result<(), Error> {
+        loop {
+            tokio::select! {
+                biased;
+
+                _ = &mut self.terminate_rx => {
+                    // If we receive a request to terminate, we are guaranteed
+                    // to not process any further requests, and to leave
+                    // immediately.
+                    self.publish_state_to_nexus().await;
+                    return Ok(());
+                },
+                // Handle messages from our own "Monitor the VMM" task.
+                request = self.rx_monitor.recv() => {
+                    use InstanceMonitorRequest::*;
+                    match request {
+                        Some(Update { state, tx }) => {
+                            let observed = ObservedPropolisState::new(
+                                self.state.instance(),
+                                &state,
+                            );
+                            let reaction = self.observe_state(&observed).await?;
```

Comment:
:+1: i agree with Greg here.

Reply:
This is mostly true -- I'll prevent the `Result` from propagating through types, since we're not using it -- but as a warning, `terminate_inner` appears to be unwrapping rather than throwing errors. It should probably be made more protected against "things going wrong during instance teardown".

Either way - updated so that the main runner returns `()`, rather than a result.